### PR TITLE
Quickgen refactor

### PR DESCRIPTION
**Pull request by rstaten**
I have moved forward with the refactor of quickgen/quickbrick in order to consolidate common code, as suggested in #180.

This PR:

- Refactors quickgen to include quickbrick functionality, allowing the user to either provide 
simspec and fibermap files or brick name to perform simulations.
  - Usage of quickgen to incorporate bricks is identical to quickbrick: i.e. 'quickbrick -b brickname' becomes 'quickgen -b brickname'.
  - Other command line options of quickbrick are included in quickgen.
- Updates quickgen unit tests to include tests for brick options.

Note: This implementation was natural following the refactor of quickgen/quickbrick code. If one script for quickgen and quickbrick use is not the preferred refactor, other options can be pursued. If quickbrick needs to remain a command line option, then a refactor that includes common code, while still allowing both quickgen and quickbrick to be given by the user, can be provided.

### CODE REVIEW

**Comment by sbailey** [comment]
I'm not sure this refactor is really accomplishing the goal of having quickgen and quickbrick share the same code.  Although they are now within the same `desisim.scripts.quickgen.main` function and share a single call to `qsim.simulate()`, the actual code is mostly split up by blocks of:
```python
if args.simspec:
    # do something
elif args.brickname:
    # do something very similar but slightly different
```
We're trying to get rid of duplicated code like
```python
    if args.simspec:
        if simspec.flavor in bright_objects:
            if args.moon_phase is None:
                qsim.atmosphere.moon.moon_phase = 0.7
            else:
                qsim.atmosphere.moon.moon_phase = args.moon_phase
        #...
```
vs.
```python
    elif args.brickname:
        if objtype in bright_objects:
            if args.moon_phase is None:
                qsim.atmosphere.moon.moon_phase = 0.7
            else:
                qsim.atmosphere.moon.moon_phase = args.moon_phase
        #...
```

I'm looking for a deeper refactor that would accomplish something like:
```python
if args.simspec:
    # read inputs from simspec file
else:
    # create inputs on the fly...

# Do stuff that is common to both quickgen and quickbrick without a bunch
# of if/then blocks for the two cases

if args.brickname:
    # write a brick file with all the calibrated spectra and exit
else:
    # write frame, sky, fluxcalib, cframe, etc.
```
Let's keep working on this...

**Comment by rstaten** [comment]
@sbailey The most recent push should encompass more of what you are looking for in the refactor. However, when running the updated unit tests, the moon phase test is failing when using brick files and I am still looking into this before a potential merge.

**Comment by rstaten** [comment]
The unit tests now all pass, and this can potentially be merged if it is the desired refactor.

File: py/desisim/scripts/quickgen.py [hunk 1]
```diff
@@ -412,10 +508,21 @@ def main(args=None):
...
             cframe_observedflux[j, i, :num_pixels] = 1e17 * output['observed_flux']
             cframe_ivar[j, i, :num_pixels] = 1e-34 * output['flux_inverse_variance']
 
+            # Fill brick arrays from the results.
+            if i==0:
+                camera = 'b'
+            elif i==1:
+                camera = 'r'
+            else:
+                camera = 'z'
```
Comment (sbailey) [inline py/desisim/scripts/quickgen.py]: I think the qsim.camera_output results are not guaranteed to be in the order b, r, z.  There must be a better way to derive their name.

File: py/desisim/scripts/quickgen.py [hunk 2]
```diff
@@ -1,109 +1,127 @@
...
+    parser.add_argument('--zrange-lrg', type=float, default=(0.5, 1.1), nargs=2, metavar='',
+                        help='minimum and maximum redshift range for LRG')
+    parser.add_argument('--zrange-bgs', type=float, default=(0.01, 0.4), nargs=2, metavar='',
+                        help='minimum and maximum redshift range for BGS')
+    parser.add_argument('--rmagrange-bgs', type=float, default=(15.0, 19.5), nargs=2, metavar='',
+                        help='Minimum and maximum BGS r-band (AB) magnitude range')
+    parser.add_argument('--sne-rfluxratiorange', type=float, default=(0.1, 1.0), nargs=2, metavar='',
+                        help='r-band flux ratio of the SNeIa spectrum relative to the galaxy')
+    parser.add_argument('--add-SNeIa', action='store_true', help='include SNeIa spectra')
+
```
Comment (sbailey) [inline py/desisim/scripts/quickgen.py]: Please add some comment documentation about:
* how to run this in "quickgen mode", outputting frame files, cframe files, etc.
* how to run this in "quickbrick mode", outputting only bricks
* which options apply to only one or the other vs. both

**Comment by rstaten** [comment]
@sbailey These changes have been implemented, so this refactor should be ready to be merged. For the record, I ran the command discussed in issue #312 on this refactored version, but did not get the reported error messages.

**Comment by sbailey** [comment]
I updated the camera name code to get it directly from the specsim results instead of mapping the mean wavelength to a name:
```python
camera = output.meta['name']
```
PR #189 changed `test_quickbrick.py`, but those tests have been moved to `test_quickgen.py` in this PR, causing a merge conflict with master.  I ported the updated test into `test_quickgen.py` in this branch.  After confirming that the Travis tests still pass, I'll take care of the command line merge to resolve the conflicts.

**Pull request merged**